### DISCUSSION TITLE
Can a Fyne app access the hardware and/or services on a mobile device

### DISCUSSION THREAD

**Discussion by serge-hulne** [fyne-io/fyne#3600] [General]
If it is possible to compile a Fyne app to run on IOS or Android, is there a mechanism whereby one could further use the hardware of the device or some of its services, like for instance use he camera or use the speech recognition functionality of the device?

Just to give a naïve example, Is it possible to record some audio from within a Fyne app by somehow calling a Swift method or a Kotlin method, respectively? What about using the camera or speech recognition or speech synthesis?

I think that Svelte Native, Rect Native, have such a mechanism, but does it exist in Fyne (or is something of that sort foreseen)?

**Comment by serge-hulne** [comment]
Perhaps I have been looking at the problem on the wrong level, here is another approach: interfacing Go and Swift:

https://medium.com/@matryer/tutorial-calling-go-code-from-swift-on-ios-and-vice-versa-with-gomobile-7925620c17a4

**Reply by andydotxyz** [reply]
> Perhaps I have been looking at the problem on the wrong level, here is another approach: interfacing Go and Swift:

Unfortunately Fyne does not support being embedded in other types of application like that article is demonstrating.

**Comment by andydotxyz** [comment]
> Just to give a naïve example, Is it possible to record some audio from within a Fyne app by somehow calling a Swift method or a Kotlin method, respectively? What about using the camera or speech recognition or speech synthesis?

You can access all system components through CGo. For iOS that means objective-C and for Android it's NDK. I don't know the specifics for the tasks you want to perform.
